File: internal/agent/compact.go
package agent

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/anthropics/anthropic-sdk-go"
	"github.com/anthropics/anthropic-sdk-go/option"
	"github.com/joho/godotenv"
)

// ── 压缩阈值 ──────────────────────────────────────────

const (
	contextLimit     = 50_000 // 触发自动摘要压缩的粗略上下文阈值（字符数）
	keepRecent       = 3      // L2 压缩时保留最近几个 tool_result 的完整内容
	persistThreshold = 30_000 // 单个工具结果超过此阈值时落盘
	maxBudget        = 200_000
)

// ── 辅助函数 ──────────────────────────────────────────

// estimateSize 粗略估算消息列表占用的上下文大小
func estimateSize(messages []anthropic.MessageParam) int {
	data, _ := json.Marshal(messages)
	return len(data)
}

// safeFilename 将 tool_use_id 转换为安全文件名
func safeFilename(value string) string {
	var sb strings.Builder
	for _, ch := range value {
		if (ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z') || (ch >= '0' && ch <= '9') || ch == '-' || ch == '_' {
			sb.WriteRune(ch)
		} else {
			sb.WriteRune('_')
		}
	}
	if sb.Len() == 0 {
		return "unknown"
	}
	return sb.String()
}

// ── 消息类型判断 ──────────────────────────────────────

// messageHasToolUse 判断 assistant 消息是否包含 tool_use block
func messageHasToolUse(m anthropic.MessageParam) bool {
	if m.Role != "assistant" {
		return false
	}
	for _, b := range m.Content {
		if b.OfToolUse != nil {
			return true
		}
	}
	return false
}

// isToolResultMessage 判断 user 消息是否包含 tool_result block
func isToolResultMessage(m anthropic.MessageParam) bool {
	if m.Role != "user" {
		return false
	}
	for _, b := range m.Content {
		if b.OfToolResult != nil {
			return true
		}
	}
	return false
}

// ── L1: snipCompact — 裁剪中间消息 ──────────────────

func snipCompact(messages []anthropic.MessageParam, maxMessages int) []anthropic.MessageParam {
	if maxMessages <= 0 {
		maxMessages = 50
	}
	if len(messages) <= maxMessages {
		return messages
	}

	keepHead := 3
	keepTail := maxMessages - keepHead
	headEnd := keepHead
	tailStart := len(messages) - keepTail

	// 如果头部最后一条是 tool_use，把紧随其后的 tool_result 也保留
	if headEnd > 0 && messageHasToolUse(messages[headEnd-1]) {
		for headEnd < len(messages) && isToolResultMessage(messages[headEnd]) {
			headEnd++
		}
	}

	// 如果尾部第一条是 tool_result，把它前面的 tool_use 也保留
	if tailStart > 0 && tailStart < len(messages) &&
		isToolResultMessage(messages[tailStart]) &&
		tailStart-1 >= 0 && messageHasToolUse(messages[tailStart-1]) {
		tailStart--
	}

	if headEnd >= tailStart {
		return messages
	}

	snipped := tailStart - headEnd
	result := make([]anthropic.MessageParam, 0, headEnd+1+len(messages)-tailStart)
	result = append(result, messages[:headEnd]...)
	result = append(result, anthropic.NewUserMessage(
		anthropic.NewTextBlock(fmt.Sprintf("[snipped %d messages]", snipped)),
	))
	result = append(result, messages[tailStart:]...)
	return result
}

// ── L2: microCompact — 压缩旧 tool_result ──────────

func microCompact(messages []anthropic.MessageParam) []anthropic.MessageParam {
	// 收集所有 tool_result block 的位置
	type toolResultRef struct {
		msgIdx   int
		blockIdx int
	}
	var refs []toolResultRef

	for i, m := range messages {
		if m.Role != "user" {
			continue
		}
		for j, b := range m.Content {
			if b.OfToolResult != nil {
				refs = append(refs, toolResultRef{i, j})
			}
		}
	}

	if len(refs) <= keepRecent {
		return messages
	}

	// 压缩较旧的 tool_result（保留最近 keepRecent 个）
	for _, ref := range refs[:len(refs)-keepRecent] {
		block := messages[ref.msgIdx].Content[ref.blockIdx].OfToolResult
		if block == nil {
			continue
		}
		content, _ := json.Marshal(block.Content)
		if len(content) > 120 {
			block.Content = []anthropic.ToolResultBlockParamContentUnion{
				{OfText: &anthropic.TextBlockParam{Text: "[Earlier tool result compacted. Re-run if needed.]"}},
			}
		}
	}
	return messages
}

// ── L3: toolResultBudget — 大结果落盘 ────────────────

// persistLargeOutput 将超大工具输出写入文件，返回包含路径和预览的轻量内容
func (rt *agentRuntime) persistLargeOutput(toolUseID string, output string) string {
	if len(output) <= persistThreshold {
		return output
	}

	os.MkdirAll(rt.config.ToolResultsDir, 0o755)
	path := filepath.Join(rt.config.ToolResultsDir, safeFilename(toolUseID)+".txt")

	// 避免重复写入
	if _, err := os.Stat(path); os.IsNotExist(err) {
		os.WriteFile(path, []byte(output), 0o644)
	}

	preview := output
	if len(preview) > 2000 {
		preview = preview[:2000]
	}

	return fmt.Sprintf("<persisted-output>\nFull output: %s\nPreview:\n%s\n</persisted-output>", path, preview)
}

func (rt *agentRuntime) toolResultBudget(messages []anthropic.MessageParam) []anthropic.MessageParam {
	if len(messages) == 0 {
		return messages
	}
	last := &messages[len(messages)-1]
	if last.Role != "user" {
		return messages
	}

	// 处理单个超大结果
	for i := range last.Content {
		block := last.Content[i].OfToolResult
		if block == nil {
			continue
		}
		content, _ := json.Marshal(block.Content)
		contentStr := string(content)
		if len(contentStr) > persistThreshold {
			toolUseID := block.ToolUseID
			persisted := rt.persistLargeOutput(toolUseID, contentStr)
			block.Content = []anthropic.ToolResultBlockParamContentUnion{
				{OfText: &anthropic.TextBlockParam{Text: persisted}},
			}
		}
	}

	return messages
}

// ── 轻量压缩组合 ──────────────────────────────────────

func (rt *agentRuntime) applyLightweightCompaction(messages []anthropic.MessageParam) []anthropic.MessageParam {
	messages = rt.toolResultBudget(messages)
	messages = snipCompact(messages, 50)
	messages = microCompact(messages)
	return messages
}

// ── L4: autoCompact — LLM 摘要压缩 ──────────────────

// writeTranscript 将对话保存为 JSONL 文件
func (rt *agentRuntime) writeTranscript(messages []anthropic.MessageParam) string {
	os.MkdirAll(rt.config.TranscriptDir, 0o755)
	path := filepath.Join(rt.config.TranscriptDir, fmt.Sprintf("transcript_%d.jsonl", time.Now().Unix()))
	f, err := os.Create(path)
	if err != nil {
		return ""
	}
	defer f.Close()

	enc := json.NewEncoder(f)
	enc.SetEscapeHTML(false)
	for _, m := range messages {
		enc.Encode(m)
	}
	return path
}

// summarizeHistory 调用模型总结历史
func (rt *agentRuntime) summarizeHistory(messages []anthropic.MessageParam) string {
	godotenv.Load()

	opts := []option.RequestOption{}
	if base := os.Getenv("ANTHROPIC_BASE_URL"); base != "" {
		opts = append(opts, option.WithBaseURL(base))
	}

	client := anthropic.NewClient(opts...)
	ctx := context.Background()

	conversation, _ := json.Marshal(messages)
	convStr := string(conversation)
	if len(convStr) > 80000 {
		convStr = convStr[:80000]
	}

	prompt := "Summarize this coding-agent conversation so work can continue.\n" +
		"Preserve: 1. current goal, 2. key findings/decisions, 3. files read/changed, " +
		"4. remaining work, 5. user constraints.\nBe compact but concrete.\n\n" + convStr

	resp, err := client.Messages.New(ctx, anthropic.MessageNewParams{
		Model:     anthropic.Model(rt.config.Model),
		Messages:  []anthropic.MessageParam{anthropic.NewUserMessage(anthropic.NewTextBlock(prompt))},
		MaxTokens: 2000,
	})
	if err != nil {
		return "(summary failed: " + err.Error() + ")"
	}

	var parts []string
	for _, block := range resp.Content {
		if tb, ok := block.AsAny().(anthropic.TextBlock); ok {
			parts = append(parts, tb.Text)
		}
	}
	summary := strings.Join(parts, "\n")
	if summary == "" {
		return "(empty summary)"
	}
	return summary
}

// compactHistory 保存 transcript 后压成一条 summary 消息
func (rt *agentRuntime) compactHistory(messages []anthropic.MessageParam) []anthropic.MessageParam {
	path := rt.writeTranscript(messages)
	if path != "" {
		rt.emitLine("[transcript saved: %s]", path)
	}
	summary := rt.summarizeHistory(messages)
	return []anthropic.MessageParam{
		anthropic.NewUserMessage(anthropic.NewTextBlock("[Compacted]\n\n" + summary)),
	}
}

// maybeCompactHistory 先做轻量压缩，仍超限则升级到 LLM 摘要
func (rt *agentRuntime) maybeCompactHistory(messages []anthropic.MessageParam) []anthropic.MessageParam {
	messages = rt.applyLightweightCompaction(messages)
	if estimateSize(messages) <= contextLimit {
		return messages
	}
	return rt.compactHistory(messages)
}

// ── Emergency: reactiveCompact ──────────────────────

// reactiveCompact API 返回 400 时的紧急压缩
func (rt *agentRuntime) reactiveCompact(messages []anthropic.MessageParam) []anthropic.MessageParam {
	path := rt.writeTranscript(messages)
	if path != "" {
		rt.emitLine("[reactive transcript saved: %s]", path)
	}
	summary := rt.summarizeHistory(messages)

	tailStart := len(messages) - 5
	if tailStart < 0 {
		tailStart = 0
	}

	// 保持 tool_use/tool_result 配对完整
	if tailStart > 0 && tailStart < len(messages) &&
		isToolResultMessage(messages[tailStart]) &&
		tailStart-1 >= 0 && messageHasToolUse(messages[tailStart-1]) {
		tailStart--
	}

	result := []anthropic.MessageParam{
		anthropic.NewUserMessage(anthropic.NewTextBlock("[Reactive compact]\n\n" + summary)),
	}
	result = append(result, messages[tailStart:]...)
	return result
}

File: s08_context_compact/compact.py
import json
import time
from pathlib import Path
from typing import Any

import anthropic
from dotenv import load_dotenv

from .constants import MODEL, WORKDIR

# 触发自动摘要压缩的粗略上下文阈值。这里用字符串长度估算，足够做教学版保护。
CONTEXT_LIMIT = 50_000

# L2 压缩时保留最近几个 tool_result 的完整内容，旧结果替换成占位符。
KEEP_RECENT = 3

# 单个工具结果超过这个阈值时，优先把完整内容落盘，只给模型保留预览。
PERSIST_THRESHOLD = 30_000

# 大输出和完整 transcript 都写到工作区内，避免写到用户主目录或不可写缓存目录。
COMPACT_DIR = WORKDIR / ".agent_state" / "compact"
TOOL_RESULTS_DIR = COMPACT_DIR / "tool_results"
TRANSCRIPT_DIR = COMPACT_DIR / "transcripts"


def estimate_size(messages: list[dict[str, Any]]) -> int:
    """粗略估算 messages 当前占用的上下文大小。"""
    return len(str(messages))


def _get_client() -> anthropic.Anthropic:
    """延迟创建 Anthropic client，避免导入 compact.py 时就要求环境变量齐全。"""
    load_dotenv(override=True)
    return anthropic.Anthropic()


def _block_type(block: Any) -> str | None:
    """兼容 dict block 和 Anthropic SDK 返回的对象 block。"""
    return (
        block.get("type") if isinstance(block, dict) else getattr(block, "type", None)
    )


def _message_has_tool_use(message: dict[str, Any]) -> bool:
    """判断 assistant 消息中是否包含 tool_use block。"""
    if message.get("role") != "assistant":
        return False

    content = message.get("content")
    if not isinstance(content, list):
        return False

    return any(_block_type(block) == "tool_use" for block in content)


def _is_tool_result_message(message: dict[str, Any]) -> bool:
    """判断 user 消息是否是工具调用结果消息。"""
    if message.get("role") != "user":
        return False

    content = message.get("content")
    if not isinstance(content, list):
        return False

    return any(
        isinstance(block, dict) and block.get("type") == "tool_result"
        for block in content
    )


def _safe_filename(value: Any) -> str:
    """把 tool_use_id 变成安全文件名，防止路径穿越或特殊字符写盘失败。"""
    text = str(value) or "unknown"
    safe = "".join(ch if ch.isalnum() or ch in "-_" else "_" for ch in text)
    return safe or "unknown"


# L1: snipCompact - 裁掉对话中间部分，只保留开头、结尾和一条占位说明。
def snip_compact(
    messages: list[dict[str, Any]], max_messages: int = 50
) -> list[dict[str, Any]]:
    """当消息数量过多时，删除中间历史，同时尽量不切断 tool_use/tool_result 配对。"""
    if len(messages) <= max_messages:
        return messages

    keep_head = 3
    keep_tail = max_messages - keep_head
    head_end = keep_head
    tail_start = len(messages) - keep_tail

    # 如果头部最后一条是 tool_use，必须把紧随其后的 tool_result 一并保留。
    if head_end > 0 and _message_has_tool_use(messages[head_end - 1]):
        while head_end < len(messages) and _is_tool_result_message(messages[head_end]):
            head_end += 1

    # 如果尾部第一条是 tool_result，必须把它前面的 assistant tool_use 一并保留。
    if (
        tail_start > 0
        and tail_start < len(messages)
        and _is_tool_result_message(messages[tail_start])
        and _message_has_tool_use(messages[tail_start - 1])
    ):
        tail_start -= 1

    if head_end >= tail_start:
        return messages

    snipped = tail_start - head_end
    return (
        messages[:head_end]
        + [{"role": "user", "content": f"[snipped {snipped} messages]"}]
        + messages[tail_start:]
    )


def collect_tool_results(
    messages: list[dict[str, Any]],
) -> list[tuple[int, int, dict[str, Any]]]:
    """收集所有 tool_result block 的位置和对象引用，供 L2/L3 复用。"""
    blocks: list[tuple[int, int, dict[str, Any]]] = []
    for message_index, message in enumerate(messages):
        if message.get("role") != "user" or not isinstance(message.get("content"), list):
            continue

        for block_index, block in enumerate(message["content"]):
            if isinstance(block, dict) and block.get("type") == "tool_result":
                blocks.append((message_index, block_index, block))

    return blocks


# L2: microCompact - 把较旧的长工具结果替换成短占位符。
def micro_compact(messages: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """保留最近 KEEP_RECENT 个 tool_result，其余较长结果压成占位文本。"""
    tool_results = collect_tool_results(messages)
    if len(tool_results) <= KEEP_RECENT:
        return messages

    for _, _, block in tool_results[:-KEEP_RECENT]:
        content = str(block.get("content", ""))
        if len(content) > 120:
            block["content"] = "[Earlier tool result compacted. Re-run if needed.]"

    return messages


# L3: toolResultBudget - 当最近一批工具结果总量太大时，把最大结果落盘。
def persist_large_output(tool_use_id: Any, output: Any) -> str:
    """把超大工具输出写入文件，并返回包含路径和预览的轻量内容。"""
    text = str(output)
    if len(text) <= PERSIST_THRESHOLD:
        return text

    TOOL_RESULTS_DIR.mkdir(parents=True, exist_ok=True)
    path = TOOL_RESULTS_DIR / f"{_safe_filename(tool_use_id)}.txt"
    if not path.exists():
        path.write_text(text)

    return (
        "<persisted-output>\n"
        f"Full output: {path}\n"
        f"Preview:\n{text[:2000]}\n"
        "</persisted-output>"
    )


def tool_result_budget(
    messages: list[dict[str, Any]], max_bytes: int = 200_000
) -> list[dict[str, Any]]:
    """落盘最后一条消息中的大工具结果，并限制这批结果的总体大小。"""
    last = messages[-1] if messages else None
    if (
        not last
        or last.get("role") != "user"
        or not isinstance(last.get("content"), list)
    ):
        return messages

    blocks = [
        (index, block)
        for index, block in enumerate(last["content"])
        if isinstance(block, dict) and block.get("type") == "tool_result"
    ]

    # 先处理单个超大结果，保证 L2 占位符替换前仍有完整内容可写入文件。
    for _, block in blocks:
        content = str(block.get("content", ""))
        if len(content) > PERSIST_THRESHOLD:
            block["content"] = persist_large_output(block.get("tool_use_id", "unknown"), content)

    total = sum(len(str(block.get("content", ""))) for _, block in blocks)
    if total <= max_bytes:
        return messages

    ranked = sorted(
        blocks, key=lambda pair: len(str(pair[1].get("content", ""))), reverse=True
    )
    for _, block in ranked:
        if total <= max_bytes:
            break

        content = str(block.get("content", ""))
        if len(content) <= PERSIST_THRESHOLD:
            continue

        block["content"] = persist_large_output(block.get("tool_use_id", "unknown"), content)
        total = sum(len(str(block.get("content", ""))) for _, block in blocks)

    return messages


def apply_lightweight_compaction(
    messages: list[dict[str, Any]],
) -> list[dict[str, Any]]:
    """执行不需要额外 LLM 调用的三层压缩：L3 -> L1 -> L2。"""
    messages = tool_result_budget(messages)
    messages = snip_compact(messages)
    return micro_compact(messages)


# L4: autoCompact - 仍然超限时，用模型把完整历史总结成一条可延续上下文。
def write_transcript(messages: list[dict[str, Any]]) -> Path:
    """把压缩前完整对话写成 jsonl，方便之后排查或恢复上下文。"""
    TRANSCRIPT_DIR.mkdir(parents=True, exist_ok=True)
    path = TRANSCRIPT_DIR / f"transcript_{int(time.time())}.jsonl"
    with path.open("w") as file:
        for message in messages:
            file.write(json.dumps(message, ensure_ascii=False, default=str) + "\n")
    return path


def summarize_history(
    messages: list[dict[str, Any]], client: anthropic.Anthropic | None = None
) -> str:
    """调用模型总结历史，保留继续工作的关键信息。"""
    conversation = json.dumps(messages, ensure_ascii=False, default=str)[:80_000]
    prompt = (
        "Summarize this coding-agent conversation so work can continue.\n"
        "Preserve: 1. current goal, 2. key findings/decisions, 3. files read/changed, "
        "4. remaining work, 5. user constraints.\nBe compact but concrete.\n\n"
        + conversation
    )
    active_client = client or _get_client()
    response = active_client.messages.create(
        model=MODEL,
        messages=[{"role": "user", "content": prompt}],
        max_tokens=2000,
    )
    summary = "\n".join(
        getattr(block, "text", "")
        for block in response.content
        if getattr(block, "type", None) == "text"
    ).strip()
    return summary or "(empty summary)"


def compact_history(messages: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """保存 transcript 后，把整段历史压成一条 summary 消息。"""
    transcript_path = write_transcript(messages)
    print(f"[transcript saved: {transcript_path}]")
    summary = summarize_history(messages)
    return [{"role": "user", "content": f"[Compacted]\n\n{summary}"}]


def maybe_compact_history(messages: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """先做轻量压缩；如果仍超过阈值，再升级到 LLM 摘要压缩。"""
    messages = apply_lightweight_compaction(messages)
    if estimate_size(messages) <= CONTEXT_LIMIT:
        return messages
    return compact_history(messages)


# Emergency: reactiveCompact - API 已经因为上下文超限失败时使用。
def reactive_compact(messages: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """保留完整 transcript、生成摘要，并带上最近几条消息继续重试。"""
    transcript_path = write_transcript(messages)
    print(f"[reactive transcript saved: {transcript_path}]")
    summary = summarize_history(messages)

    tail_start = max(0, len(messages) - 5)
    if (
        tail_start > 0
        and tail_start < len(messages)
        and _is_tool_result_message(messages[tail_start])
        and _message_has_tool_use(messages[tail_start - 1])
    ):
        tail_start -= 1

    return [
        {"role": "user", "content": f"[Reactive compact]\n\n{summary}"},
        *messages[tail_start:],
    ]
